xla/service/cpu/tests/onednn_memory_util_test.cc

```diff
@@ -0,0 +1,109 @@
+limitations under the License.
+==============================================================================*/
+
+#if defined(INTEL_MKL) && defined(ENABLE_ONEDNN_V3)
+
+#include "xla/service/cpu/onednn_memory_util.h"
+
+#include <string>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/IR/Argument.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Type.h"
+#include "llvm/IR/Value.h"
+#include "xla/service/llvm_ir/llvm_util.h"
+#include "xla/shape.h"
+#include "xla/shape_util.h"
+#include "xla/test.h"
+#include "xla/tests/filecheck.h"
+#include "xla/tsl/lib/core/status_test_util.h"
+#include "tsl/platform/test.h"
+
+namespace xla {
+namespace cpu {
+namespace {
+
+class MemoryUtilTest : public ::testing::Test,
+                       public ::testing::WithParamInterface<
+                           std::tuple<PrimitiveType, std::vector<int64_t>>> {
+ protected:
+  constexpr static const char* pad_test_pattern_ = R"(
+    CHECK: %[[mref0:[0-9]+]] = insertvalue
+    CHECK: %[[mref1:[0-9]+]] = insertvalue
+    CHECK-SAME: [[arr:\[12 x i64\]]], [[arr]] } %[[mref0]], i64 255, 3
+  )";
+};
+
+TEST_P(MemoryUtilTest, VerifyPadTest) {
+  PrimitiveType dtype = std::get<0>(GetParam());
+  std::string filecheck_input;
+  llvm::LLVMContext context = llvm::LLVMContext();
+  llvm::IRBuilder builder(context);
+  llvm::raw_string_ostream ostream(filecheck_input);
+  llvm::Module module("MemoryUtilPad", context);
+
+  llvm::FunctionType* function_type = llvm::FunctionType::get(
+      llvm::Type::getVoidTy(context), {builder.getPtrTy()}, false);
+  llvm::Function* function = llvm::Function::Create(
+      function_type, llvm::Function::LinkageTypes::ExternalLinkage,
+      "memory_util_pad_test", module);
+  llvm::BasicBlock* bb = llvm::BasicBlock::Create(context, "BB", function);
+  builder.SetInsertPoint(bb);
+
+  Shape shape = ShapeUtil::MakeShape(dtype, std::get<1>(GetParam()));
+  llvm::Argument* ptr = function->getArg(0);
+  llvm::Type* type = llvm_ir::PrimitiveTypeToIrType(dtype, &module);
+
+  if (shape.IsArray()) {
+    for (auto dim : LayoutUtil::MinorToMajor(shape)) {
+      type = llvm::ArrayType::get(type, shape.dimensions(dim));
+    }
+  }
+
+  llvm_ir::IrArray ir_array(ptr, type, shape);
+  auto alloca = GetAllocaAndEmitMemrefInfo(builder, ir_array);
+  alloca.EmitLifetimeEnd();
+  ostream << module;
+
+  absl::StatusOr<bool> match = RunFileCheck(filecheck_input, pad_test_pattern_);
+  TF_ASSERT_OK(match.status());
+  EXPECT_TRUE(match.value());
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    MemoryUtilTestSuite, MemoryUtilTest,
+    ::testing::Combine(::testing::ValuesIn({S8, S16, BF16, F16, F32}),
+                       ::testing::Values(std::vector<int64_t>({30}),
+                                         std::vector<int64_t>({30, 40}),
+                                         std::vector<int64_t>({30, 40, 50}))),
```

Comment:
I don't think we need to input many dtypes/shapes since we don't check for them in the filecheck. (Or we could parameterize on just the shape, and add a check line to look for the actual sizes.)